pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -341,13 +356,19 @@ func (r *ReconcileConstraintTemplate) handleUpdate(
 			log.Error(err, "failed to report constraint template ingestion duration")
 		}
 		err := r.reportErrorOnCTStatus("ingest_error", "Could not ingest Rego", ct, err)
+		r.tracker.CancelTemplate(unversionedCT) // Don't track templates that failed compilation
 		return reconcile.Result{}, err
 	}
 
 	if err := r.metrics.reportIngestDuration(metrics.ActiveStatus, time.Since(beginCompile)); err != nil {
 		log.Error(err, "failed to report constraint template ingestion duration")
 	}
 
+	// Mark for readiness tracking
+	t := r.tracker.For(gvkConstraintTemplate)
+	t.Observe(unversionedCT)
```

Comment:
Is there something like `setDataExpectations` for constraints and templates as well? After a quick look I don't see any roster gathering code.

Reply:
Yes, it's in [`trackConstraintTemplates`](https://github.com/open-policy-agent/gatekeeper/blob/a22c7e5638745b22d0ef9ea31e00a08d47f150ba/pkg/readiness/ready_tracker.go#L218).

Reply:
Thanks! Is there a race condition putting it in Run(), given that other controllers and the server (who is executing this health check) are also starting up in parallel?

Reply:
`CancelTemplates` is synchronized with `Run` via the `t.ready` channel because it depends on `constraintTemplates` being initialized. Other than that, objectTrackers are internally synchronized. Observations are allowed to occur before expectations. There is a matter of avoiding new constraint expectations once the original roster is locked in - I'm thinking about different approaches to address this now.

Reply:
The above is resolved now.